### ISSUE TITLE
DB issue when updating to 5.7.3

### ISSUE THREAD

**Issue by das-peter** [pimcore/pimcore#4233]
# Bug Report
Running the update from 5.7.2 to 5.7.3 leads to following error:

>In AbstractMySQLDriver.php line 106:
>
>  An exception occurred while executing 'UPDATE `object_metadata_29` mt
>                           INNER JOIN `object_relations_29` rl
>                           ON  mt.o_id = rl.src_id
>                               AND mt.dest_id = rl.dest_id
>                               AND mt.type = rl.type
>                               AND mt.fieldname = rl.fieldname
>                               AND mt.ownertype = rl.ownertype
>                               AND mt.ownername = rl.ownername
>                               AND mt.position = rl.position
>                           SET mt.index = rl.index':
> 
>   SQLSTATE[42000]: Syntax error or access violation: 1071 Specified key was too long; max key length is 1024 bytes

I think the reason for this is that the primary key of metadata table can surpass the index size limit of mysql. The special thing about this is that the key size limit doesn't seem what one would expect.
If you're using the [innodb_large_prefix](https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_large_prefix) you might think that the key size limit is 3072 bytes as documented [here](https://dev.mysql.com/doc/refman/5.7/en/innodb-restrictions.html).

Question is why does mysql complain about the key exceeding 1024 bytes then?
Best bet is in this discussion going on stackexchange about that: https://dba.stackexchange.com/questions/217793/mysql-refuses-update-with-1071-specified-key-was-too-long-max-key-length-is
TL;DR could be because of following change in MySQL > 5.7.23: 
https://github.com/mysql/mysql-server/commit/1217a6c91e84c3645a8975e5620bb49289f9f9a6

But how to fix something like this?

### Approach one: Shrink the index size

So this would mean that we'd need to shrink the max size of the primary index of the table.

I've tried to look into the table and figure out the best way to achieve that.
First question is if we really need to support utf8mb4 for all columns.
* `fieldname` is a mysql identifier according to the [documentation](https://dev.mysql.com/doc/refman/8.0/en/identifiers.html) this supports:
    >Permitted characters in quoted identifiers include the full Unicode Basic Multilingual Plane (BMP), except U+0000:
    >   ASCII: U+0001 .. U+007F
    >   Extended: U+0080 .. U+FFFF 

   If I understand correctly https://en.wikipedia.org/wiki/Plane_(Unicode)#Basic_Multilingual_Plane is utf8 in mysql and doesn't need utf8mb4.
    Further if I see that correctly we've got a 70 char limit for that in other places - I guess we can / should use the same limit here.
* `ownertype` is already limited to the enum of: `'object','fieldcollection','localizedfield','objectbrick'`
So this is good old ascii. No need to swing the utf8mb4 hammer.
* `type`: If that is somehow related to ownertype I'd say we can use utf8 instead utf8mb4.
* `column`: Sounds pretty much like it has an identifier behind it - so utf8 should do too.
* `owernname`, `position`: Let's just hope we don't need utf8mb4 because most of this seems to be identifers of some sort :P
* `data`: Well that's some sort of text, and not part of the index anyways. So we'll keep that at utf8mb4 - no touchy.

All that given I end up with a max-index that still exceeds the mentioned 1024 limit:

Field|Char-Limit|Charset-Bytes|Index-Bytes
-----|----------|-------------|-----------
o_id|11|1|11
dest_id|11|1|11
type|50|1|50
fieldname|70|3|210
column|190|3|570
data|0|4|0
ownertype|15|1|15
ownername|70|1|70
position|70|3|210
index|11|1|11

Max-Index-Size: 1158 bytes

So we nee to go down by quite a bit to be within the limit - which we could do limiting the key size:

Field|Char-Limit|Key-Limit|Charset-Bytes|Index-Bytes
-----|----------|---------|-------------|-----------
o_id|11|11|1|11
dest_id|11|11|1|11
type|50|50|1|50
fieldname|70|62|3|186
column|190|161|3|483
data|0|0|4|0
ownertype|15|15|1|15
ownername|70|62|1|62
position|70|62|3|186
index|11|11|1|11

Max-Index-size: 1015 bytes

That could do the trick but the question is if this is suitable for something that is the a primary key...

However, the SQL for a metadata table with these properties would look like this:
```sql
CREATE TABLE `object_metadata_29` (
  `o_id` int(11) NOT NULL DEFAULT '0',
  `dest_id` int(11) NOT NULL DEFAULT '0',
  `type` varchar(50) CHARACTER SET ascii NOT NULL,
  `fieldname` varchar(70) CHARACTER SET utf8 NOT NULL DEFAULT '',
  `column` varchar(190) CHARACTER SET utf8 NOT NULL,
  `data` text,
  `ownertype` enum('object','fieldcollection','localizedfield','objectbrick') CHARACTER SET ascii NOT NULL DEFAULT 'object',
  `ownername` varchar(70) CHARACTER SET utf8 NOT NULL DEFAULT '',
  `position` varchar(70) CHARACTER SET utf8 NOT NULL DEFAULT '0',
  `index` int(11) unsigned NOT NULL DEFAULT '0',
  PRIMARY KEY (`o_id`,`dest_id`,`type`,`fieldname`(62),`column`(161),`ownertype`,`ownername`(62),`position`(62),`index`),
  KEY `o_id` (`o_id`),
  KEY `dest_id` (`dest_id`),
  KEY `fieldname` (`fieldname`),
  KEY `column` (`column`),
  KEY `ownertype` (`ownertype`),
  KEY `ownername` (`ownername`),
  KEY `position` (`position`),
  KEY `index` (`index`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 ROW_FORMAT=COMPRESSED;
```

### Approach Two: Avoid hitting the root cause

This approach is way less involved but also introduces "special knowledge" as we don't solve the root cause and hence risk that the issue is hit by other queries / usages.

In my tests following update query works without hitting the key limit issue:
```sql
UPDATE `object_metadata_29` mt
  SET mt.index = (SELECT rl.index FROM `object_relations_29` rl
    WHERE  mt.o_id = rl.src_id
    AND mt.dest_id = rl.dest_id
    AND mt.type = rl.type
    AND mt.fieldname = rl.fieldname
    AND mt.ownertype = rl.ownertype
    AND mt.ownername = rl.ownername
    AND mt.position = rl.position)
```

So we could just adjust the update code instead tampering with the table - but we risk that we hit the same issue somewhere else and someone has to dig again into this to figure out what the heck went wrong.

### Finally

Both approaches are less than ideal and someone with way more pimcore experience is required to properly evaluate which one is the most suitable.

**Comment by brusch** [comment]
Thank you very much for your bug report & the corresponding PR! 
I'm running MariaDB 10.1.37 and unfortunately I'm not able to reproduce the problem. 

But as far as i understood this problem is only related to multi-table updates (when using `JOIN` in your `UPDATE` statement) so your proposed fix in #4234 would solve this issue, as the schema definition is basically fine. Even with `uft8mb4` the current key length is `1838` so by far lower than the maximum of `3072`.

**Comment by das-peter** [comment]
Thanks for the swift response.
I'm afraid the issue is very much related to what MySQL Version / flavor one uses.
We're running https://hub.docker.com/_/percona (percona:5.7) which currently is at 5.7.25-28.

And percona is still quite close to MySQL which makes it more likely to be hit by https://github.com/mysql/mysql-server/commit/1217a6c91e84c3645a8975e5620bb49289f9f9a6

So I guess for now the adjusted query is less invasive than the 1024 byte key length fix.
